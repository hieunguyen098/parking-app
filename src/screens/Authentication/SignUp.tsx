import { StyleSheet, View, Image, Pressable } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import LargeButton from '../../components/Buttons/LargeButton';
import FieldInput from './components/FieldInput';
import DateInput from './components/DateInput';
import Select from './components/Select';
import * as ImagePicker from 'expo-image-picker';
import { GlobalStyles } from '../../constants/style';

const genders = [
    {
        id: 0,
        label: 'Nam',
        value: 1,
    },
    {
        id: 1,
        label: 'Nữ',
        value: 0,
    },
    {
        id: 2,
        label: 'Khác',
        value: -1,
    },
];
const SignUp = () => {
    const navigation: any = useNavigation();
    const nextStep = () => {
        navigation.navigate('CreatePassword');
    };
    const [image, setImage] = useState('');

    const pickImage = async () => {
        // No permissions request is necessary for launching the image library
        let result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.Images,
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        if (!result.canceled) {
            setImage(result.assets[0].uri);
        }
    };

    const takePhoto = async () => {
        let result = await ImagePicker.launchCameraAsync({
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        if (!result.canceled) {
            setImage(result.assets[0].uri);
        }
    };

    return (
        <View style={styles.container}>
            <View>
                <Pressable onPress={takePhoto}>
                    <Image
                        source={image ? { uri: image } : require('../../../assets/images/default-avt-0.png')}
                        style={styles.avatar}
                    />
                </Pressable>
                <Pressable style={styles.editAvatarBtn} onPress={pickImage}>
                    <Image
                        style={{ width: '100%', height: '100%' }}
                        source={require('../../../assets/images/edit-icon.png')}
                    />
                </Pressable>
            </View>
            <View style={styles.form}>
                <FieldInput placeHolder="Họ tên" iconPosition="none" keyboardType="text" />
                <DateInput
                    placeHolder="Ngày sinh"
                    source={require('../../../assets/images/calendar-icon.png')}
                    iconPosition="right"
                    keyboardType="text"
                />
                <FieldInput
                    placeHolder="Email"
                    source={require('../../../assets/images/mail-icon.png')}
                    iconPosition="right"
                    keyboardType="email-address"
                />
                <Select data={genders} />
            </View>
            <LargeButton onPress={nextStep} type="primary" title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default SignUp;

const styles = StyleSheet.create({
    container: {
        backgroundColor: '#fff',
        padding: 20,
        flex: 1,
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    continueButton: {
        marginBottom: 24,
        borderRadius: 100,
    },
    avatar: {
        width: 200,
        height: 200,
        borderRadius: 100,
    },
    form: {
        width: '100%',
    },
    editAvatarBtn: {
        position: 'absolute',
        right: 0,
        bottom: 0,
        backgroundColor: GlobalStyles.colors.primaryOrange,
        padding: 5,
        width: 40,
        height: 40,
        borderRadius: 5,
    },
});
