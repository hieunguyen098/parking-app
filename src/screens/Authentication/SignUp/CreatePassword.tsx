import { StyleSheet, Text, View } from 'react-native';
import React, { useState } from 'react';
import LargeButton from '../../../components/Buttons/LargeButton';
import { useNavigation } from '@react-navigation/native';
import PasswordInput from '../../../components/PasswordInput/PasswordInput';
import styles from '../styles';
import { signup } from '../../../services/auth';
import { useSelector } from 'react-redux';

const CreatePassword = () => {
    const navigation: any = useNavigation();
    const [pin, setPin] = useState('');
    const [pinConfirm, setPinConfirm] = useState('');
    const signupForm = useSelector((state: any) => state.auth.signupForm);
    const { phone } = useSelector((state: any) => state.auth.user)
    const [showError, setShowError] = useState(false)
    const onNext = async () => {
        if (pin !== pinConfirm) {
            setShowError(true)
            return
        }
        const data = await signup({
            pin,
            phone,
            ...signupForm,
        });
        if (data.returnCode > 0) {
            navigation.navigate('PhoneVerification');
        }
        else {
            console.log(data)
        }
    };
    return (
        <View style={styles.container}>
            <View style={styles.group}>
                <Text style={styles.description}>Nhập mật khẩu của bạn</Text>
                <PasswordInput warning={{ show: showError, message: "Mật khẩu không khớp." }} onChange={() => setShowError(false)} value={pin} setValue={setPin} length={6} />
                <Text style={styles.description}>Nhập lại mật khẩu</Text>
                <PasswordInput warning={{ show: showError, message: "Mật khẩu không khớp." }} onChange={() => setShowError(false)} value={pinConfirm} setValue={setPinConfirm} length={6} />
            </View>
            <LargeButton onPress={() => onNext()} title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default CreatePassword;
