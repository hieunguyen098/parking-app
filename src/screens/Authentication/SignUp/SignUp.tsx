import { View, Image, Pressable } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import LargeButton from '../../../components/Buttons/LargeButton';
import * as ImagePicker from 'expo-image-picker';
import styles from '../styles';
import FieldInput from '../../../components/FieldInput';
import DateInput from '../../../components/DateInput';
import Select from '../../../components/Select';
import { useDispatch, useSelector } from 'react-redux';
import { authActions } from '../../../../store/slices/authSlice';
import { storage } from '../../../firebase';
import { ref, uploadBytes, getDownloadURL } from 'firebase/storage';

const genders = [
    {
        id: 0,
        label: 'Nam',
        value: 'Male',
    },
    {
        id: 1,
        label: 'Nữ',
        value: 'Female',
    },
    {
        id: 2,
        label: 'Khác',
        value: 'Indefinite',
    },
];

const SignUp = () => {
    const navigation: any = useNavigation();
    const dispatch = useDispatch();
    const nextStep = async () => {
        dispatch(
            authActions.updateSignupForm({
                fullname,
                birthday,
                email,
                image_url,
                gender,
            }),
        );
        navigation.navigate('CreatePassword');
    };
    const user = useSelector((state: any) => state.auth.user);
    const [fullname, setFullname] = useState('');
    const [birthday, setBirthday] = useState('');
    const [email, setEmail] = useState('');
    const [image_url, setImageUrl] = useState('');
    const [gender, setGender] = useState(genders[0].value);

    const pickImage = async () => {
        // No permissions request is necessary for launching the image library
        const result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.Images,
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        await handleImagePicked(result);
    };

    const takePhoto = async () => {
        const result = await ImagePicker.launchCameraAsync({
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        await handleImagePicked(result);
    };

    const handleImagePicked = async (pickerResult: any) => {
        try {
            console.log('Uploading');
            if (!pickerResult.canceled) {
                await uploadImageAsync(pickerResult.assets[0].uri);
            }
        } catch (e) {
            console.log(e);
            alert('Upload failed, sorry :(');
        } finally {
            console.log('Uploaded');
        }
    };

    async function uploadImageAsync(uri: string) {
        // Why are we using XMLHttpRequest? See:
        // https://github.com/expo/expo/issues/2402#issuecomment-443726662
        const blob = await new Promise((resolve, reject) => {
            const xhr = new XMLHttpRequest();
            xhr.onload = function () {
                resolve(xhr.response);
            };
            xhr.onerror = function (e) {
                console.log(e);
                reject(new TypeError('Network request failed'));
            };
            xhr.responseType = 'blob';
            xhr.open('GET', uri, true);
            xhr.send(null);
        });

        const fileRef = ref(storage, `avatar/${Date.now()}-${user.phone}`);
        // @ts-ignore
        await uploadBytes(fileRef, blob).then(async (snapshot) => {
            console.log('Uploaded a blob or file!');
            await getDownloadURL(snapshot.ref).then((downloadURL) => {
                console.log('File available at', downloadURL);
                setImageUrl(downloadURL);
            });
        });

        // We're done with the blob, close and release it
        // @ts-ignore
        blob.close();
    }

    return (
        <View style={styles.container}>
            <View>
                <Pressable onPress={takePhoto}>
                    <Image
                        source={image_url ? { uri: image_url } : require('../../../../assets/images/default-avt-0.png')}
                        style={styles.avatar}
                    />
                </Pressable>
                <Pressable style={styles.editAvatarBtn} onPress={pickImage}>
                    <Image
                        style={{ width: '100%', height: '100%' }}
                        source={require('../../../../assets/images/edit-icon.png')}
                    />
                </Pressable>
            </View>
            <View style={styles.form}>
                <FieldInput
                    value={fullname}
                    setValue={setFullname}
                    placeHolder="Họ tên"
                    iconPosition="none"
                    keyboardType="text"
                />
                <DateInput
                    value={birthday}
                    setValue={setBirthday}
                    placeHolder="Ngày sinh"
                    source={require('../../../../assets/images/calendar-icon.png')}
                    iconPosition="right"
                    keyboardType="text"
                />
                <FieldInput
                    value={email}
                    setValue={setEmail}
                    placeHolder="Email"
                    source={require('../../../../assets/images/mail-icon.png')}
                    iconPosition="right"
                    keyboardType="email-address"
                />
                <Select value={gender} setValue={setGender} data={genders} />
            </View>
            <LargeButton onPress={nextStep} title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default SignUp;
