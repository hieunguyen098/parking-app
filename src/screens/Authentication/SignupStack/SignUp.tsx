import { StyleSheet, View, Image, Pressable } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import LargeButton from '../../../components/Buttons/LargeButton';
import FieldInput from '../components/FieldInput';
import DateInput from '../components/DateInput';
import Select from '../components/Select';
import * as ImagePicker from 'expo-image-picker';
import { GlobalStyles } from '../../../constants/style';
import styles from '../styles';

const genders = [
    {
        id: 0,
        label: 'Nam',
        value: 1,
    },
    {
        id: 1,
        label: 'Nữ',
        value: 0,
    },
    {
        id: 2,
        label: 'Khác',
        value: -1,
    },
];
const SignUp = () => {
    const navigation: any = useNavigation();
    const nextStep = () => {
        navigation.navigate('CreatePassword');
    };
    const [image, setImage] = useState('');

    const pickImage = async () => {
        // No permissions request is necessary for launching the image library
        let result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.Images,
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        if (!result.canceled) {
            setImage(result.assets[0].uri);
        }
    };

    const takePhoto = async () => {
        let result = await ImagePicker.launchCameraAsync({
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        if (!result.canceled) {
            setImage(result.assets[0].uri);
        }
    };

    return (
        <View style={styles.container}>
            <View>
                <Pressable onPress={takePhoto}>
                    <Image
                        source={image ? { uri: image } : require('../../../../assets/images/default-avt-0.png')}
                        style={styles.avatar}
                    />
                </Pressable>
                <Pressable style={styles.editAvatarBtn} onPress={pickImage}>
                    <Image
                        style={{ width: '100%', height: '100%' }}
                        source={require('../../../../assets/images/edit-icon.png')}
                    />
                </Pressable>
            </View>
            <View style={styles.form}>
                <FieldInput placeHolder="Họ tên" iconPosition="none" keyboardType="text" />
                <DateInput
                    placeHolder="Ngày sinh"
                    source={require('../../../../assets/images/calendar-icon.png')}
                    iconPosition="right"
                    keyboardType="text"
                />
                <FieldInput
                    placeHolder="Email"
                    source={require('../../../../assets/images/mail-icon.png')}
                    iconPosition="right"
                    keyboardType="email-address"
                />
                <Select data={genders} />
            </View>
            <LargeButton onPress={nextStep} title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default SignUp;
