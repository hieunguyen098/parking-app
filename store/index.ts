import { configureStore } from '@reduxjs/toolkit';
import { authReducer } from './slices/authSlice';
import { generalReducer } from './slices/generalSlice';

export const store = configureStore({
    reducer: {
        auth: authReducer,
        general: generalReducer,
    },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
