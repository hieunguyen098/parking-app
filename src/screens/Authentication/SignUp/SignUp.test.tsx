import React from 'react';
import { render } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';

import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { store } from '../../../../store';
import SignUp from './SignUp';

jest.useFakeTimers();
test('test sign up', () => {
    const queryClient = new QueryClient();
    const component = (
        <QueryClientProvider client={queryClient}>
            <Provider store={store}>
                <NavigationContainer>
                    <SignUp />
                </NavigationContainer>
            </Provider>
        </QueryClientProvider>
    );

    const tree = render(component);
    expect(tree).toMatchSnapshot();
});
