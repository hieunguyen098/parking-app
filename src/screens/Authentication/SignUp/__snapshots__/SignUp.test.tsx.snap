// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test sign up 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "space-between",
      "padding": 20,
    }
  }
>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={1}
        style={
          Object {
            "borderRadius": 100,
            "height": 150,
            "width": 150,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FF953A",
          "borderRadius": 5,
          "bottom": 0,
          "height": 30,
          "padding": 5,
          "position": "absolute",
          "right": 0,
          "width": 30,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FBFBFB",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginVertical": 10,
              "paddingHorizontal": 20,
            },
            undefined,
          ]
        }
      >
        <TextInput
          keyboardType="text"
          onChangeText={[Function]}
          placeholder="Họ tên"
          style={
            Object {
              "flex": 1,
              "fontSize": 16,
              "paddingVertical": 20,
            }
          }
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FBFBFB",
          "borderRadius": 20,
          "flexDirection": "row",
          "marginVertical": 10,
          "paddingHorizontal": 20,
        }
      }
    >
      <TextInput
        keyboardType="text"
        onPressIn={[Function]}
        placeholder="Ngày sinh"
        showSoftInputOnFocus={false}
        style={
          Object {
            "flex": 1,
            "fontSize": 16,
            "paddingVertical": 20,
          }
        }
        value=""
      />
      <Image
        source={1}
        style={
          Object {
            "height": 25,
            "width": 25,
          }
        }
      />
    </View>
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FBFBFB",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginVertical": 10,
              "paddingHorizontal": 20,
            },
            undefined,
          ]
        }
      >
        <TextInput
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Email"
          style={
            Object {
              "flex": 1,
              "fontSize": 16,
              "paddingVertical": 20,
            }
          }
          value=""
        />
        <Image
          source={1}
          style={
            Object {
              "height": 25,
              "width": 25,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FBFBFB",
          "borderRadius": 20,
          "flexDirection": "row",
          "marginVertical": 10,
          "paddingLeft": 15,
          "paddingRight": 12,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "fontSize": 16,
            "paddingVertical": 35,
          }
        }
      >
        <RNCPicker
          items={
            Array [
              Object {
                "label": "Nam",
                "testID": undefined,
                "textColor": undefined,
                "value": "Male",
              },
              Object {
                "label": "Nữ",
                "testID": undefined,
                "textColor": undefined,
                "value": "Female",
              },
              Object {
                "label": "Khác",
                "testID": undefined,
                "textColor": undefined,
                "value": "Indefinite",
              },
            ]
          }
          numberOfLines={1}
          onChange={[Function]}
          selectedIndex={0}
          style={
            Array [
              Object {
                "height": 216,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#FF953A",
          "borderRadius": 8,
          "paddingHorizontal": 12,
          "paddingVertical": 18,
          "width": "100%",
        },
        Object {
          "backgroundColor": "#FF953A",
        },
        Object {
          "borderRadius": 100,
          "marginBottom": 24,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 18,
            "fontWeight": "600",
            "textAlign": "center",
          },
          Object {
            "color": "#fff",
          },
          undefined,
        ]
      }
    >
      Tiếp tục
    </Text>
  </View>
</View>
`;
